File: packages/go/enums.go
package rdk

// The closed value sets accepted by the QoreChain rdk module. These mirror the
// on-chain rdk module exactly. The string values are the wire values the chain
// expects -- do not localize or re-case them.

// SettlementParadigm is how a rollup settles to the Main Chain.
type SettlementParadigm string

const (
	SettlementOptimistic SettlementParadigm = "optimistic"
	SettlementZK         SettlementParadigm = "zk"
	SettlementBased      SettlementParadigm = "based"
	SettlementSovereign  SettlementParadigm = "sovereign"
)

// SettlementParadigms enumerates every valid SettlementParadigm.
var SettlementParadigms = []SettlementParadigm{
	SettlementOptimistic,
	SettlementZK,
	SettlementBased,
	SettlementSovereign,
}

// SequencerMode is who orders the rollup's transactions.
type SequencerMode string

const (
	SequencerDedicated SequencerMode = "dedicated"
	SequencerShared    SequencerMode = "shared"
	SequencerBased     SequencerMode = "based"
)

// SequencerModes enumerates every valid SequencerMode.
var SequencerModes = []SequencerMode{
	SequencerDedicated,
	SequencerShared,
	SequencerBased,
}

// ProofSystem is the proof a settlement batch carries.
type ProofSystem string

const (
	ProofFraud ProofSystem = "fraud"
	ProofSnark ProofSystem = "snark"
	ProofStark ProofSystem = "stark"
	ProofNone  ProofSystem = "none"
)

// ProofSystems enumerates every valid ProofSystem.
var ProofSystems = []ProofSystem{
	ProofFraud,
	ProofSnark,
	ProofStark,
	ProofNone,
}

// DABackend is where rollup data is made available.
type DABackend string

const (
	DANative   DABackend = "native"
	DACelestia DABackend = "celestia"
	DABoth     DABackend = "both"
)

// DABackends enumerates every valid DABackend.
var DABackends = []DABackend{
	DANative,
	DACelestia,
	DABoth,
}

// GasModel is the fee model the rollup charges.
type GasModel string

const (
	GasStandard   GasModel = "standard"
	GasEIP1559    GasModel = "eip1559"
	GasFlat       GasModel = "flat"
	GasSubsidized GasModel = "subsidized"
)

// GasModels enumerates every valid GasModel.
var GasModels = []GasModel{
	GasStandard,
	GasEIP1559,
	GasFlat,
	GasSubsidized,
}

// VmType is the execution environment the rollup exposes.
//
//   - "evm"    -- Ethereum/Solidity.
//   - "native" -- the QoreChain Native runtime (Wasm smart contracts).
//   - "svm"    -- the Solana VM.
//   - "custom" -- an application-defined VM.
//
// "cosmwasm" is accepted as a legacy alias of "native" and is what both map to
// on the wire (the network, explorer, and dashboard use "cosmwasm"). Do not
// localize or re-case these values.
type VmType string

const (
	VmEVM      VmType = "evm"
	VmNative   VmType = "native"
	VmSVM      VmType = "svm"
	VmCustom   VmType = "custom"
	VmCosmWasm VmType = "cosmwasm"
)

// VmTypes enumerates the advertised VmType values. "native" is the QoreChain
// Native runtime; the "cosmwasm" legacy alias is accepted (see IsVMType) but
// deliberately kept out of the advertised list.
var VmTypes = []VmType{
	VmEVM,
	VmNative,
	VmSVM,
	VmCustom,
}

// acceptedVMTypes is the set of VM types IsVMType accepts: the four advertised
// values plus the "cosmwasm" legacy alias.
var acceptedVMTypes = map[string]struct{}{
	string(VmEVM):      {},
	string(VmNative):   {},
	string(VmSVM):      {},
	string(VmCustom):   {},
	string(VmCosmWasm): {},
}

// IsVMType reports whether value is a VM type the RDK accepts (the four
// advertised values plus the "cosmwasm" legacy alias).
func IsVMType(value string) bool {
	_, ok := acceptedVMTypes[value]
	return ok
}

// VMTypeWireValue returns the on-chain wire value for a VM type. The QoreChain
// Native runtime ("native") is transmitted as "cosmwasm" for consistency with
// the network, explorer, and dashboard; all other values pass through
// unchanged. The wire value is never "native".
func VMTypeWireValue(vmType string) string {
	if vmType == string(VmNative) {
		return string(VmCosmWasm)
	}
	return vmType
}

// VMTypeLabel returns a human-readable label for a VM type. The Wasm runtime
// (native/cosmwasm) reads as "QoreChain Native".
func VMTypeLabel(vmType string) string {
	switch vmType {
	case string(VmNative), string(VmCosmWasm):
		return "QoreChain Native"
	case string(VmEVM):
		return "EVM"
	case string(VmSVM):
		return "SVM"
	case string(VmCustom):
		return "Custom"
	default:
		return vmType
	}
}

// RollupStatus is a rollup lifecycle state.
type RollupStatus string

const (
	RollupPending RollupStatus = "pending"
	RollupActive  RollupStatus = "active"
	RollupPaused  RollupStatus = "paused"
	RollupStopped RollupStatus = "stopped"
)

// RollupStatuses enumerates every valid RollupStatus.
var RollupStatuses = []RollupStatus{
	RollupPending,
	RollupActive,
	RollupPaused,
	RollupStopped,
}

// BatchStatus is a settlement-batch lifecycle state.
type BatchStatus string

const (
	BatchSubmitted  BatchStatus = "submitted"
	BatchChallenged BatchStatus = "challenged"
	BatchFinalized  BatchStatus = "finalized"
	BatchRejected   BatchStatus = "rejected"
)

// BatchStatuses enumerates every valid BatchStatus.
var BatchStatuses = []BatchStatus{
	BatchSubmitted,
	BatchChallenged,
	BatchFinalized,
	BatchRejected,
}

// Profile is one of the five preset profiles.
type Profile string

const (
	ProfileDefi       Profile = "defi"
	ProfileGaming     Profile = "gaming"
	ProfileNFT        Profile = "nft"
	ProfileEnterprise Profile = "enterprise"
	ProfileCustom     Profile = "custom"
)

// Profiles enumerates every preset Profile.
var Profiles = []Profile{
	ProfileDefi,
	ProfileGaming,
	ProfileNFT,
	ProfileEnterprise,
	ProfileCustom,
}

File: packages/py/src/qorrdk/enums.py
"""Closed value sets accepted by the QoreChain ``rdk`` module.

These mirror the on-chain ``rdk`` module exactly. The string values are the wire
values the chain expects -- do not localize or re-case them. Each enum is a
``str`` enum so members compare equal to their wire string.
"""

from __future__ import annotations

from enum import Enum


class SettlementParadigm(str, Enum):
    """How a rollup settles to the Main Chain."""

    OPTIMISTIC = "optimistic"
    ZK = "zk"
    BASED = "based"
    SOVEREIGN = "sovereign"


class SequencerMode(str, Enum):
    """Who orders the rollup's transactions."""

    DEDICATED = "dedicated"
    SHARED = "shared"
    BASED = "based"


class ProofSystem(str, Enum):
    """The proof a settlement batch carries."""

    FRAUD = "fraud"
    SNARK = "snark"
    STARK = "stark"
    NONE = "none"


class DABackend(str, Enum):
    """Where rollup data is made available."""

    NATIVE = "native"
    CELESTIA = "celestia"
    BOTH = "both"


class GasModel(str, Enum):
    """The fee model the rollup charges."""

    STANDARD = "standard"
    EIP1559 = "eip1559"
    FLAT = "flat"
    SUBSIDIZED = "subsidized"


class VmType(str, Enum):
    """The execution environment the rollup exposes.

    - ``EVM`` -- Ethereum/Solidity.
    - ``NATIVE`` -- the QoreChain Native runtime (Wasm smart contracts).
    - ``SVM`` -- the Solana VM.
    - ``CUSTOM`` -- an application-defined VM.

    ``COSMWASM`` is accepted as a legacy alias of ``NATIVE`` and is what both map
    to on the wire (the network, explorer, and dashboard use ``cosmwasm``). It is
    not part of the advertised :data:`VM_TYPES`.
    """

    EVM = "evm"
    NATIVE = "native"
    SVM = "svm"
    CUSTOM = "custom"
    #: Legacy alias of :attr:`NATIVE`; the on-chain wire value both resolve to.
    COSMWASM = "cosmwasm"


class RollupStatus(str, Enum):
    """Rollup lifecycle states."""

    PENDING = "pending"
    ACTIVE = "active"
    PAUSED = "paused"
    STOPPED = "stopped"


class BatchStatus(str, Enum):
    """Settlement-batch lifecycle states."""

    SUBMITTED = "submitted"
    CHALLENGED = "challenged"
    FINALIZED = "finalized"
    REJECTED = "rejected"


class ProfileName(str, Enum):
    """The five preset profiles."""

    DEFI = "defi"
    GAMING = "gaming"
    NFT = "nft"
    ENTERPRISE = "enterprise"
    CUSTOM = "custom"


SETTLEMENT_PARADIGMS: tuple[str, ...] = tuple(m.value for m in SettlementParadigm)
SEQUENCER_MODES: tuple[str, ...] = tuple(m.value for m in SequencerMode)
PROOF_SYSTEMS: tuple[str, ...] = tuple(m.value for m in ProofSystem)
DA_BACKENDS: tuple[str, ...] = tuple(m.value for m in DABackend)
GAS_MODELS: tuple[str, ...] = tuple(m.value for m in GasModel)
#: The advertised VM types (``native`` is the QoreChain Native runtime). The
#: ``cosmwasm`` legacy alias is accepted by :func:`is_vm_type` but not advertised.
VM_TYPES: tuple[str, ...] = ("evm", "native", "svm", "custom")
ROLLUP_STATUSES: tuple[str, ...] = tuple(m.value for m in RollupStatus)
BATCH_STATUSES: tuple[str, ...] = tuple(m.value for m in BatchStatus)
PROFILE_NAMES: tuple[str, ...] = tuple(m.value for m in ProfileName)

_ACCEPTED_VM_TYPES: frozenset[str] = frozenset((*VM_TYPES, "cosmwasm"))


def is_vm_type(value: object) -> bool:
    """Whether ``value`` is a VM type the RDK accepts (includes the ``cosmwasm`` alias)."""
    return getattr(value, "value", value) in _ACCEPTED_VM_TYPES


def vm_type_wire_value(vm_type: object) -> str:
    """The on-chain wire value for a VM type.

    The QoreChain Native runtime (``native``) is transmitted as ``cosmwasm`` for
    consistency with the network, explorer, and dashboard; all other values pass
    through unchanged.
    """
    value = str(getattr(vm_type, "value", vm_type))
    return "cosmwasm" if value == "native" else value


def vm_type_label(vm_type: object) -> str:
    """A human-readable label for a VM type (the Wasm runtime reads as QoreChain Native)."""
    value = str(getattr(vm_type, "value", vm_type))
    if value in ("native", "cosmwasm"):
        return "QoreChain Native"
    if value == "evm":
        return "EVM"
    if value == "svm":
        return "SVM"
    if value == "custom":
        return "Custom"
    return value


__all__ = [
    "SettlementParadigm",
    "SequencerMode",
    "ProofSystem",
    "DABackend",
    "GasModel",
    "VmType",
    "RollupStatus",
    "BatchStatus",
    "ProfileName",
    "SETTLEMENT_PARADIGMS",
    "SEQUENCER_MODES",
    "PROOF_SYSTEMS",
    "DA_BACKENDS",
    "GAS_MODELS",
    "VM_TYPES",
    "ROLLUP_STATUSES",
    "BATCH_STATUSES",
    "PROFILE_NAMES",
    "is_vm_type",
    "vm_type_wire_value",
    "vm_type_label",
]
